### ISSUE TITLE
Improve tableName support

### ISSUE THREAD

**Issue by modomodo** [yiisoft/yii2-gii#170]
Hi, 
I'm currently working on building a Microsoft Dynamics Nav web portal using Yii2 and the Gii generator

### What steps will reproduce the problem?
Upon entering table names used in the database backend, for example: 

> Test & Movies$Transcription

The Gii Model generator throws a validation error, stating that:
`Only word characters, and optionally spaces, an asterisk and/or a dot are allowed`

### What's expected?
I would love support for various table naming schemes instead of being limited to the ones defined in the Gii generator: 
```php
/^([\w ]+\.)?([\w\* ]+)$/
```
### What do you get instead?
![image](https://cloud.githubusercontent.com/assets/5679726/17125045/db293588-52f8-11e6-8405-6f9726a53a55.png)

### Additional info

| Q                | A
| ---------------- | ---
| Yii vesion       | 2.0.9
| PHP version      | 5.6.22
| Operating system | Windows 10 64 bit
| Database | SQL Server 2014

**Comment by modomodo** [comment]
Out of curiosity, why are tableNames limited in the Gii Model Generator?

**Comment by samdark** [comment]
Gii generates relations as methods. There are limitations on characters we can use within PHP so it's a good idea to apply such limitations right away.

These validators are here:

https://github.com/yiisoft/yii2-gii/blob/master/generators/model/Generator.php#L76

It would be great if you'll adjust validation and see how it goes with such table names.

**Comment by modomodo** [comment]
Will try it out... And update this issue on the results. I'm quite new to regular expressions, so it would take me some time to create one that supports the table name: **Test & Movies$Transcription**

**Comment by samdark** [comment]
Just use `.*` for tests which would allow anything. I'm more concerned about PHP code generated than about how it's validated.

**Comment by modomodo** [comment]
@samdark Sorry that it took me long to get a test case, had a hectic tuesday and morning. I edited the regex as you suggested, which meant that there were no validation errors:
![image](https://cloud.githubusercontent.com/assets/5679726/17174108/b182d710-5407-11e6-97d6-533f847e9ad4.png)

However, this caused the model generator command which creates the model, view and controller to timeout with the following error:
```php
PHP Fatal Error – yii\base\ErrorException

Maximum execution time of 30 seconds exceeded
1. in [redacted]vendor\yiisoft\yii2\db\Command.php at line 842
        $this->prepare(true);
 
        $token = $rawSql;
        try {
            Yii::beginProfile($token, 'yii\db\Command::query');
 
            $this->pdoStatement->execute();
 
            if ($method === '') {
                $result = new DataReader($this);
            } else {
                if ($fetchMode === null) {
                    $fetchMode = $this->fetchMode;
                }
                $result = call_user_func_array([$this->pdoStatement, $method], (array) $fetchMode);
                $this->pdoStatement->closeCursor();
            }

```

I'm going to try and implement a more specific regex supporting the use case needed

**Comment by samdark** [comment]
It's not the regex that causes timeout.

**Comment by modomodo** [comment]
Okay, so isn't there any way of supporting table names with characters such as: `$ &` ?

How could one drill down into what's causing the timeout.

Because, strangely enough, using a custom model created from scratch works within Yii2 and the model performs all SQL supported operations through Yii2 without any problems or errors. Such as this model:

```php
<?php

namespace common\models;

use Yii;

class Submission extends \yii\db\ActiveRecord
{
    /**
     * @inheritdoc
     */
    public static function tableName()
    {
        return 'Ritman Test$AssignmentSubmission';
    }
 public function getAssignments()  
	{
        return $this->hasOne(Assignments::className(), ['AssignmentID' => 'AssignmentID'])->from(Assignments::tableName());
    }
```

**Comment by samdark** [comment]
It's Gii which times out, not the AR itself. Based on your stack trace I guess it happens while querying schema.

**Comment by modomodo** [comment]
So, is this an issue to be looked into?

**Comment by samdark** [comment]
Yes, but it's low priority since there aren't many developers using such names for schema. If you need it I'd recommend to look into it yourself and then pull request a solution.

**Comment by modomodo** [comment]
I do need it. I also found that using `tablePrefix` helps out with the table names. However, using prefixes with table names that have white spaces result in the following weird error:

`
SQLSTATE[42000]: [Microsoft][ODBC Driver 11 for SQL Server]Syntax error, permission violation, or other nonspecific error
Failed to prepare SQL: SELECT COUNT(*) FROM {{%[Example [Table]}}]
`
Actual table in model using prefix: `"{{Example Table}}"`
Yii2 is trying to Alias the specified name in the model

**Comment by modomodo** [comment]
Any tips on what parts of Yii are relevant?

**Comment by samdark** [comment]
I think Gii extension in the first place which doesn't work with these.
